DATA WRANGLING AND CLEANING:

First, we will import necessary libraries.


In [1]:
import pandas as pd
import numpy as np

Then, we will scrape the Kaggle webpage for the .csv file.


In [2]:
#Produces error. File must be pre-cleaned.
#url = 'https://www.kaggle.com/kandij/diabetes-dataset?select=diabetes2.csv'
#df = pd.read_csv(url)
#df.head()

There is an error in line 7, where a comma was missing. Instead of accessing the file online, the file is downloaded and patched. The fixed file, named diabetes2.csv, is uploaded to GitHub.

In [18]:
file = 'diabetes2.csv'
df = pd.read_csv(file)
df.reset_index()
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

Various other checks are performed, such as .info() and .describe()

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


All columns appear numeric at first glance. However, Outcome is our target, categorical variable represented as 0 for non-diabetic and 1 for diabetic.  

In [20]:
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [63]:
df.shape

(768, 9)

Looks like no null values are present. It doesn't make sense that some values for blood pressure, skin thickness, or BMI are zero.

In [21]:
df[df['BloodPressure'] == 0]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
7             10      115              0              0        0  35.3   
15             7      100              0              0        0  30.0   
49             7      105              0              0        0   0.0   
60             2       84              0              0        0   0.0   
78             0      131              0              0        0  43.2   
81             2       74              0              0        0   0.0   
172            2       87              0             23        0  28.9   
193           11      135              0              0        0  52.3   
222            7      119              0              0        0  25.2   
261            3      141              0              0        0  30.0   
266            0      138              0              0        0  36.3   
269            2      146              0              0        0  27.5   
300            0      167              0              0        0  32.3   
332            1      180              0              0        0  43.3   
336            0      117              0              0        0  33.8   
347            3      116              0              0        0  23.5   
357           13      129              0             30        0  39.9   
426            0       94              0              0        0   0.0   
430            2       99              0              0        0  22.2   
435            0      141              0              0        0  42.4   
453            2      119              0              0        0  19.6   
468            8      120              0              0        0  30.0   
484            0      145              0              0        0  44.2   
494            3       80              0              0        0   0.0   
522            6      114              0              0        0   0.0   
533            6       91              0              0        0  29.8   
535            4      132              0              0        0  32.9   
589            0       73              0              0        0  21.1   
601            6       96              0              0        0  23.7   
604            4      183              0              0        0  28.4   
619            0      119              0              0        0  32.4   
643            4       90              0              0        0  28.0   
697            0       99              0              0        0  25.0   
703            2      129              0              0        0  38.5   
706           10      115              0              0        0   0.0   

     DiabetesPedigreeFunction  Age  Outcome  
7                       0.134   29        0  
15                      0.484   32        1  
49                      0.305   24        0  
60                      0.304   21        0  
78                      0.270   26        1  
81                      0.102   22        0  
172                     0.773   25        0  
193                     0.578   40        1  
222                     0.209   37        0  
261                     0.761   27        1  
266                     0.933   25        1  
269                     0.240   28        1  
300                     0.839   30        1  
332                     0.282   41        1  
336                     0.932   44        0  
347                     0.187   23        0  
357                     0.569   44        1  
426                     0.256   25        0  
430                     0.108   23        0  
435                     0.205   29        1  
453                     0.832   72        0  
468                     0.183   38        1  
484                     0.630   31        1  
494                     0.174   22        0  
522                     0.189   26        0  
533                     0.501   31        0  
535                     0.302   23        1  
589                     0.342   25        0  
601   

For blood pressure, that's a lot of rows with 0 blood pressure. Rather than removing the rows, imputing the mean of blood pressure without zeros seems to be the way to go. To impute the zeros, zeros have to be switched with 'NaN' first.

In [25]:
df.loc[df.BloodPressure == 0, 'BloodPressure'] = np.nan
df[df['BloodPressure'] == 0]

Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

Now that the zeroes from blood pressure are gone, it's time to impute the mean of blood pressure into NaNs.

In [43]:
BP_mean = df['BloodPressure'].mean()
#print(BP_mean)
df.fillna(BP_mean, axis=1, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    float64
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(3), int64(6)
memory usage: 54.1 KB


Now to handle skin thickness:

In [44]:
df[df['SkinThickness'] == 0]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
2              8      183      64.000000              0        0  23.3   
5              5      116      74.000000              0        0  25.6   
7             10      115      72.405738              0        0  35.3   
9              8      125      96.000000              0        0   0.0   
10             4      110      92.000000              0        0  37.6   
..           ...      ...            ...            ...      ...   ...   
757            0      123      72.000000              0        0  36.3   
758            1      106      76.000000              0        0  37.5   
759            6      190      92.000000              0        0  35.5   
762            9       89      62.000000              0        0  22.5   
766            1      126      60.000000              0        0  30.1   

     DiabetesPedigreeFunction  Age  Outcome  
2                       0.672   32        1  
5                       0.201   30        0  
7                       0.134   29        0  
9                       0.232   54        1  
10                      0.191   30        0  
..                        ...  ...      ...  
757                     0.258   52        1  
758                     0.197   26        0  
759                     0.278   66        1  
762                     0.142   33        0  
766                     0.349   47        1  

[227 rows x 9 columns]

In [45]:
df.loc[df.SkinThickness == 0, 'SkinThickness'] = np.nan
df[df['SkinThickness'] == 0]

Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

In [49]:
ST_mean = df['SkinThickness'].mean()
#print(ST_mean)
df.fillna(ST_mean, axis=1, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    float64
 3   SkinThickness             768 non-null    float64
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(4), int64(5)
memory usage: 54.1 KB


BMI is next:

In [50]:
df[df['BMI'] == 0]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin  BMI  \
9              8      125      96.000000       29.15342        0  0.0   
49             7      105      72.405738       29.15342        0  0.0   
60             2       84      72.405738       29.15342        0  0.0   
81             2       74      72.405738       29.15342        0  0.0   
145            0      102      75.000000       23.00000        0  0.0   
371            0      118      64.000000       23.00000       89  0.0   
426            0       94      72.405738       29.15342        0  0.0   
494            3       80      72.405738       29.15342        0  0.0   
522            6      114      72.405738       29.15342        0  0.0   
684            5      136      82.000000       29.15342        0  0.0   
706           10      115      72.405738       29.15342        0  0.0   

     DiabetesPedigreeFunction  Age  Outcome  
9                       0.232   54        1  
49                      0.305   24        0  
60                      0.304   21        0  
81                      0.102   22        0  
145                     0.572   21        0  
371                     1.731   21        0  
426                     0.256   25        0  
494                     0.174   22        0  
522                     0.189   26        0  
684                     0.640   69        0  
706                     0.261   30        1

In [53]:
df.loc[df.BMI == 0, 'BMI'] = np.nan
df[df['BMI'] == 0]

Empty DataFrame
Columns: [Pregnancies, Glucose, BloodPressure, SkinThickness, Insulin, BMI, DiabetesPedigreeFunction, Age, Outcome]
Index: []

In [55]:
BMI_mean = df['BMI'].mean()
print(BMI_mean)
df.fillna(BMI_mean, axis=1, inplace=True)
df.info()

32.45746367239099
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    float64
 3   SkinThickness             768 non-null    float64
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(4), int64(5)
memory usage: 54.1 KB


Row values are taken care of, but what about columns? Do any columns need to be dropped?

In [56]:
df.value_counts()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin  BMI        DiabetesPedigreeFunction  Age  Outcome
0            57       60.000000      29.15342       0        21.700000  0.735                     67   0          1
             67       76.000000      29.15342       0        45.300000  0.194                     46   0          1
5            103      108.000000     37.00000       0        39.200000  0.305                     65   0          1
             104      74.000000      29.15342       0        28.800000  0.153                     48   0          1
             105      72.000000      29.00000       325      36.900000  0.159                     28   0          1
                                                                                                                 ..
2            84       72.405738      29.15342       0        32.457464  0.304                     21   0          1
             85       65.000000      29.15342       0        39.600000  0.930

Let's take a closer look at Insulin.

In [57]:
df['Insulin'].value_counts()

0      374
105     11
140      9
130      9
120      8
      ... 
193      1
191      1
188      1
184      1
846      1
Name: Insulin, Length: 186, dtype: int64

In [60]:
df['Insulin'].unique()

array([  0,  94, 168,  88, 543, 846, 175, 230,  83,  96, 235, 146, 115,
       140, 110, 245,  54, 192, 207,  70, 240,  82,  36,  23, 300, 342,
       304, 142, 128,  38, 100,  90, 270,  71, 125, 176,  48,  64, 228,
        76, 220,  40, 152,  18, 135, 495,  37,  51,  99, 145, 225,  49,
        50,  92, 325,  63, 284, 119, 204, 155, 485,  53, 114, 105, 285,
       156,  78, 130,  55,  58, 160, 210, 318,  44, 190, 280,  87, 271,
       129, 120, 478,  56,  32, 744, 370,  45, 194, 680, 402, 258, 375,
       150,  67,  57, 116, 278, 122, 545,  75,  74, 182, 360, 215, 184,
        42, 132, 148, 180, 205,  85, 231,  29,  68,  52, 255, 171,  73,
       108,  43, 167, 249, 293,  66, 465,  89, 158,  84,  72,  59,  81,
       196, 415, 275, 165, 579, 310,  61, 474, 170, 277,  60,  14,  95,
       237, 191, 328, 250, 480, 265, 193,  79,  86, 326, 188, 106,  65,
       166, 274,  77, 126, 330, 600, 185,  25,  41, 272, 321, 144,  15,
       183,  91,  46, 440, 159, 540, 200, 335, 387,  22, 291, 39

In [61]:
df[df['Insulin'] == 0]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148      72.405738       35.00000        0  33.6   
1              1       85      66.000000       29.00000        0  26.6   
2              8      183      64.000000       29.15342        0  23.3   
5              5      116      74.000000       29.15342        0  25.6   
7             10      115      72.405738       29.15342        0  35.3   
..           ...      ...            ...            ...      ...   ...   
761            9      170      74.000000       31.00000        0  44.0   
762            9       89      62.000000       29.15342        0  22.5   
764            2      122      70.000000       27.00000        0  36.8   
766            1      126      60.000000       29.15342        0  30.1   
767            1       93      70.000000       31.00000        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
5                       0.201   30        0  
7                       0.134   29        0  
..                        ...  ...      ...  
761                     0.403   43        1  
762                     0.142   33        0  
764                     0.340   27        0  
766                     0.349   47        1  
767                     0.315   23        0  

[374 rows x 9 columns]

374 rows of zeros is a lot, but the patient is most likely not taking insulin if the value is zero. There is enough numerical data in this column to keep it.

In [62]:
df.head(15)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin        BMI  \
0             6      148      72.405738       35.00000        0  33.600000   
1             1       85      66.000000       29.00000        0  26.600000   
2             8      183      64.000000       29.15342        0  23.300000   
3             1       89      66.000000       23.00000       94  28.100000   
4             0      137      40.000000       35.00000      168  43.100000   
5             5      116      74.000000       29.15342        0  25.600000   
6             3       78      50.000000       32.00000       88  31.000000   
7            10      115      72.405738       29.15342        0  35.300000   
8             2      197      70.000000       45.00000      543  30.500000   
9             8      125      96.000000       29.15342        0  32.457464   
10            4      110      92.000000       29.15342        0  37.600000   
11           10      168      74.000000       29.15342        0  38.000000   
12           10      139      80.000000       29.15342        0  27.100000   
13            1      189      60.000000       23.00000      846  30.100000   
14            5      166      72.000000       19.00000      175  25.800000   

    DiabetesPedigreeFunction  Age  Outcome  
0                      0.627   50        1  
1                      0.351   31        0  
2                      0.672   32        1  
3                      0.167   21        0  
4                      2.288   33        1  
5                      0.201   30        0  
6                      0.248   26        1  
7                      0.134   29        0  
8                      0.158   53        1  
9                      0.232   54        1  
10                     0.191   30        0  
11                     0.537   34        1  
12                     1.441   57        0  
13                     0.398   59        1  
14                     0.587   51        1

This concludes the data wrangling and cleaning steps. Next will be exploratory data analysis.